### PR TITLE
Fix use-after-free of `ConditionVariable` in `ResourceLoader`

### PR DESCRIPTION
**Pull request by Ratio2**
This fix crash on android.
Crash reproduced in 4.1.2, 4.1.4 and 4.2.2.
Crash logcat with:
```
scons platform=android target=template_debug dev_build=yes CCFLAGS=-fsanitize=hwaddress LINKFLAGS=-fsanitize=hwaddress
```
[android-crash-3.txt](https://github.com/user-attachments/files/16392993/android-crash-3.txt)

### CODE REVIEW

**Comment by RandomShaper** [comment]
Thing that would help:
- Further info about under which situations the issue happens and so why such a fix is needed.
- A minimal reproduction project.
- Check if #94169 fixes it already.

**Comment by Ratio2** [comment]
> Thing that would help
The crash sometimes happens when loading a certain number of scenes before the game starts. The game engine should not crash under any circumstances, so this fix is ​​necessary.

I can't provide a minimal example, because it would take too much time. It's already taken a lot of time to catch this bug, fix it and test it.

See the android-crash-3.txt file in the first post, it shows that another thread destroys the object while it is in use. MR https://github.com/godotengine/godot/pull/94169 is unlikely to help, since the code for creating and deleting condition_variable remains unchanged.

Detailed description of incorrect work with condition_variable:
1. We start loading the resource.
2. Someone wants to get this resource, but it is not ready. The condition_variable is created and waiting occurs.
3. After the loading is complete, we call notify_all and immediately delete the condition_variable. But we have no right to do this, because another thread is using the object.

**Comment by RandomShaper** [comment]
I see. So, one thread may still be running `ConditionVariable::wait()` when another one decides to kill the cond var. And, even worse, multiple awaiters could end up trying to do the same. That's definitely a defect.

Now, your fix has a few code style issues, such as adding a seemingly multi-purposed helper class. The way `ResourceLoader` is written, which matches all the other lower level components in Godot overall is to favor explicitness. I'd suggest you to add instead an `awaiters_count` field to `LoadTask` and mend the logic as needed, in-place.

PS: I didn't need an MRP in the end. (Thank goodness! 😅)

**Comment by Ratio2** [comment]
> Now, your fix has a few code style issues, such as adding a seemingly multi-purposed helper class. The way `ResourceLoader` is written, which matches all the other lower level components in Godot overall is to favor explicitness. I'd suggest you to add instead an `awaiters_count` field to `LoadTask` and mend the logic as needed, in-place.

Done

File: core/io/resource_loader.cpp [hunk 1]
```diff
@@ -716,15 +715,20 @@ Ref<Resource> ResourceLoader::_load_complete_inner(LoadToken &p_load_token, Erro
 						DEV_ASSERT(wtp_task_err == OK);
 						thread_load_mutex.lock();
 					}
-				} else {
+				} else if (load_task.need_wait) {
 					// Loading thread is main or user thread.
 					if (!load_task.cond_var) {
 						load_task.cond_var = memnew(ConditionVariable);
 					}
+					++load_task.awaiters_count;
```
Comment (RandomShaper) [inline core/io/resource_loader.cpp:723]:
```suggestion
					load_task.awaiters_count++;
```

File: core/io/resource_loader.cpp [hunk 2]
```diff
@@ -716,15 +715,20 @@ Ref<Resource> ResourceLoader::_load_complete_inner(LoadToken &p_load_token, Erro
...
 					do {
 						load_task.cond_var->wait(p_thread_load_lock);
 						DEV_ASSERT(thread_load_tasks.has(p_load_token.local_path) && p_load_token.get_reference_count());
-					} while (load_task.cond_var);
+					} while (load_task.need_wait);
+					if (!--load_task.awaiters_count) {
```
Comment (RandomShaper) [inline core/io/resource_loader.cpp:728]:
```suggestion
					load_task.awaiters_count--;
					if (load_task.awaiters_count == 0) {
```

**Review by RandomShaper — CHANGES_REQUESTED** [review]
Looks great. Just little style changes and we're good to go.

**Comment by AThousandShips** [comment]
Please squash your commits into one, see [here](https://docs.godotengine.org/en/latest/contributing/workflow/pr_workflow.html#the-interactive-rebase)

**Comment by Ratio2** [comment]
> Please squash your commits into one

Done

**Review by RandomShaper — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Comment by RandomShaper** [comment]
Cherry-pick for 4.3 included in #96606.

**Pull request merged**